### PR TITLE
Enable playing video from local asset.

### PR DESCRIPTION
**Pull request by szakarias**

### CODE REVIEW

File: packages/video_player/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayerPlugin.java [hunk 1]
```diff
@@ -31,12 +32,45 @@
     private final EventChannel eventChannel;
     private boolean isInitialized = false;
 
+
+    VideoPlayer(
+        final EventChannel eventChannel,
```
Comment (mravn-google) [inline packages/video_player/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayerPlugin.java]: None of these parameters need to be `final`. Same in the other constructor.

File: packages/video_player/example/lib/main.dart [hunk 2]
```diff
@@ -196,6 +216,28 @@ class _PlayerLifeCycleState extends State<PlayerLifeCycle> {
   Widget build(BuildContext context) {
     return widget.childBuilder(context, controller);
   }
+
+  VideoPlayerController createVideoPlayerController();
+}
+
+class _NetworkPlayerLifeCycleState extends _PlayerLifeCycleState {
+
+  _NetworkPlayerLifeCycleState();
```
Comment (mravn-google) [inline packages/video_player/example/lib/main.dart]: Redundant

File: packages/video_player/example/lib/main.dart [hunk 3]
```diff
@@ -196,6 +216,28 @@ class _PlayerLifeCycleState extends State<PlayerLifeCycle> {
...
+
+  @override
+  VideoPlayerController createVideoPlayerController() {
+    return new VideoPlayerController.network(widget.dataSource);
+  }
+}
+
+class _AssetPlayerLifeCycleState extends _PlayerLifeCycleState {
+
+  _AssetPlayerLifeCycleState();
```
Comment (mravn-google) [inline packages/video_player/example/lib/main.dart]: Redundant

File: packages/video_player/ios/Classes/VideoPlayerPlugin.m [hunk 4]
```diff
@@ -264,16 +271,19 @@ + (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
...
-                                           messenger:[registrar messenger]];
+                                           messenger:[registrar messenger]
+                                           registrar:registrar];
   [registrar addMethodCallDelegate:instance channel:channel];
 }
 
 - (instancetype)initWithRegistry:(NSObject<FlutterTextureRegistry>*)registry
-                       messenger:(NSObject<FlutterBinaryMessenger>*)messenger {
+                       messenger:(NSObject<FlutterBinaryMessenger>*)messenger
+                       registrar:(NSObject<FlutterPluginRegistrar>*)registrar {
```
Comment (mravn-google) [inline packages/video_player/ios/Classes/VideoPlayerPlugin.m]: Maybe reduce parameters to just the `registrar`, then access the `FlutterBinaryMessenger` and the `FlutterTextureRegistry` from there.

File: packages/video_player/ios/Classes/VideoPlayerPlugin.m [hunk 5]
```diff
@@ -287,10 +297,18 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
...
-    NSString* dataSource = argsMap[@"dataSource"];
     FLTFrameUpdater* frameUpdater = [[FLTFrameUpdater alloc] initWithRegistry:_registry];
-    FLTVideoPlayer* player = [[FLTVideoPlayer alloc] initWithURL:[NSURL URLWithString:dataSource]
-                                                    frameUpdater:frameUpdater];
+    NSString* dataSource = argsMap[@"asset"];
+    FLTVideoPlayer* player;
+    if (dataSource) {
+      NSString* assetPath = [_registrar lookupKeyForAsset:dataSource];
+      player = [[FLTVideoPlayer alloc] initWithAsset:assetPath
+                                        frameUpdater:frameUpdater];
```
Comment (mravn-google) [inline packages/video_player/ios/Classes/VideoPlayerPlugin.m]: Reformat?

File: packages/video_player/lib/video_player.dart [hunk 6]
```diff
@@ -132,22 +132,27 @@ class VideoPlayerValue {
 /// After [dispose] all further calls are ignored.
 class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
   int _textureId;
-  final String uri;
   Timer timer;
   bool isDisposed = false;
   Completer<Null> _creatingCompleter;
   StreamSubscription<dynamic> _eventSubscription;
   _VideoAppLifeCycleObserver _lifeCycleObserver;
+  dynamic _createMethodArg;
```
Comment (mravn-google) [inline packages/video_player/lib/video_player.dart]: It is a bit sad that we need to carry this map around for the lifetime of the controller, given that it is only used during initialization. What about having `final String dataSource` and `final bool isNetwork` fields instead? Those may be useful for other purposes. Error reporting on playback failure should be very different for a network datasource compared to an asset one.

**Comment by sethladd** [comment]
FYI: this PR came up in a Reddit thread here: https://www.reddit.com/r/FlutterDev/comments/87qqn2/has_anyone_used_video_in_an_app/ :)

**Comment by fmatosqg** [comment]
@szakarias 

I gave a hand to @osorito who is using windows and could try version 0.3.0 but not on version 0.4.0. 
```
Using hardware rendering with device Android SDK built for x86. If you get graphics artifacts, consider enabling software rendering with "--enable-softwa
re-rendering".
Launching lib/main.dart on Android SDK built for x86 in debug mode...
Initializing gradle...
C:\Users\Omar Osorio\AppData\Roaming\Pub\Cache\hosted\pub.dartlang.org\video_player-0.4.0\android\src\main\java\io\flutter\plugins\videoplayer\VideoPlaye
rPlugin.java:244: error: cannot find symbol
                    registrar.lookupKeyForAsset(
                             ^
  symbol:   method lookupKeyForAsset(String,String)
  location: variable registrar of type Registrar
C:\Users\Omar Osorio\AppData\Roaming\Pub\Cache\hosted\pub.dartlang.org\video_player-0.4.0\android\src\main\java\io\flutter\plugins\videoplayer\VideoPlaye
rPlugin.java:247: error: cannot find symbol
                assetLookupKey = registrar.lookupKeyForAsset((String) call.argument("asset"));
                                          ^
  symbol:   method lookupKeyForAsset(String)
  location: variable registrar of type Registrar
2 errors

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':video_player:compileDebugJavaWithJavac'.
> Compilation failed; see the compiler error output for details.
```

Both versions work for me on OS X, although I'm using `Flutter (Channel dev, v0.2.2-pre.2, on Mac OS X 10.12.6 16G29, locale en-AU)`

Dart code is different, we tried to follow from the examples published on github. Only difference in dart from v 0.3.0 to 0.4.0 was 

I'm not sure this is the place for this kind of problem reporting but I couldn't open a new issue.

Also I'm not sure how to track the code for this change, my educated guess is that he's not bringing flutter.jar with his gradle configuration. That said I don't know how to find the commit for version 0.3.0 / 0.4.0 and have a look at all the changes.

I've also heard from other people in gitter having problems in java / swift compiling when importing different packages, is there a documentation where I can understand the configuration for gradle / pods (?) and how they bring flutter code into native projects so build tools find them?

**Comment by osorito** [comment]
This is the change that made it work under version 0.3.0 

class _NetworkPlayerLifeCycleState extends _PlayerLifeCycleState {
  @override
  VideoPlayerController createVideoPlayerController() {
    return new VideoPlayerController(widget.dataSource);
  }
}

class _AssetPlayerLifeCycleState extends _PlayerLifeCycleState {
  @override
  VideoPlayerController createVideoPlayerController() {
    //return new VideoPlayerController.asset(widget.dataSource);
    return new VideoPlayerController(widget.dataSource);
  }
}

**Comment by osorito** [comment]
This is my flutter doctor result 

C:\Users\Omar Osorio\AndroidStudioProjects\flutter_video_example>flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel beta, v0.2.3, on Microsoft Windows [Version 6.1.7601], locale en-US)
[√] Android toolchain - develop for Android devices (Android SDK 27.0.3)
[√] Android Studio (version 3.1)
[√] Android Studio (version 3.2)
[√] IntelliJ IDEA Community Edition (version 2017.3)
[√] Connected devices (1 available)

• No issues found!


Since i'm using Android Studio 3.1  in my graddle propperties 

#Fri Jun 23 08:50:38 CEST 2017
distributionBase=GRADLE_USER_HOME
distributionPath=wrapper/dists
zipStoreBase=GRADLE_USER_HOME
zipStorePath=wrapper/dists
distributionUrl=https\://services.gradle.org/distributions/gradle-4.4-all.zip

And my build graddle 

dependencies {
        classpath 'com.android.tools.build:gradle:3.1.0'
    }

**Comment by fmatosqg** [comment]
I found the problem, 
Java interface `Registrar` created method `lookupKeyForAsset` that is still not available on beta build. Once beta catches up with dev I'd expect it to work.

That's completely understandable since it's both backwards compatible and you guys are entitled to API changes while in beta. But please understand that this strongly affects the perception of stability of the build system. 

It would be nice to have a note added to this plugin such as "Use flutter 0.2.x or later".

**Comment by mravn-google** [comment]
Thanks for diving into this. Yes, the `video_player` plugin should have had its Flutter SDK dependency bumped. I'll do that immediately.

**Comment by mravn-google** [comment]
https://github.com/flutter/plugins/pull/474

**Comment by masudsarker093** [comment]
Using Chewie video player can I play video from local storage? For example, I want to download video from network and save to device and want to play in apps by  Chewie video player. Is it possible? 

async {
                final _openFile = await OpenFile.open(_filePath);

Want to open this video file into Chewie video player.

**Pull request merged**